org/mozilla/jss/nss/SSL.java

```diff
@@ -471,10 +471,38 @@ public static int EnableAlertLogging(SSLFDProxy fd) {
      */
     public static native int ConfigSyncTrustManagerCertAuthCallback(SSLFDProxy fd);
 
+    /**
+     * Use an asynchronous bad certificate handler which allows us to approve
+     * rejected certificates. This allows us to bypass the hostname check
+     * failure caused by the Java socket having no knowledge of the hostname
+     * we use for certificate validation; no HostnameVerifier is passed in.
+     * As a result, NSS has no value for the hostname and validation will fail.
+     *
+     * Note: This does NOT work for server-side connections.
+     *
+     * See also: SSL_BadCertHook in /usr/include/nss3/ssl.h and
+     *           JSSL_SSLFDAsyncBadCertCallback in jss/nss/SSLFDProxy.c
+     */
+    public static native int ConfigAsyncBadCertCallback(SSLFDProxy fd);
```

Comment:
When should we use `ConfigAsyncBadCertCallback` as opposed to `ConfigSyncBadCertCallback`?

Reply:
Synchronous `BadCertCallback` is good when the validation is short. It also is the only method that works on both client-side and server-side connections.

Asynchronous `BadCertCallback` is good when the validation takes a long time and/or invokes other network requests. It works **only** on client-side connections (which this is).

Because ours is short, I went with the former.

The same reasoning relates to `CertAuthHandler`, which is already present in JSSEngine. With user-supplied `TrustManager`s, we check whether or not we can be async and prefer that (issuing a `DelegatedTask` to actually run this validation).

In the future, since we might include other BadCert callback hooks, we might use async, so I added this too.